### ISSUE TITLE
Query about inference stage

### ISSUE THREAD

**Issue by SharkSpicy-NLP** [UCSB-NLP-Chang/KVLink#5]
Thanks for your great work.

During inference, `model(input_ids=input_ids, attention_mask=attention_mask_4d)` performs full attention computation before applying the 4D attention mask. How can we reduce the time-to-first-token (TTFT) in this process?

<img width="784" height="538" alt="Image" src="https://github.com/user-attachments/assets/ff8a58ca-a223-4409-8e2c-0d2adbb065e5" />

**Comment by KimperYang** [comment]
Thank you for the comment. Yes, in this setup, the model performs full attention before applying the attention mask. The attention mask here corresponds to the left panel in Figure 2. This design enables efficient training and evaluation, as it avoids using a “for loop” to compute and concatenate separate KV caches for every forward pass.

In actual deployment, however, each document’s cache is pre-computed and stored. These caches are then concatenated and used with the attention mask as shown in the right panel of Figure 2 (which is equivalent to the left one). In this scenario, since each document no longer needs to be re-encoded, the total token forward time (TTFT) is significantly reduced.
